src/partition/ReceivedPartition.cpp

```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+        vertexListRequests.push_back(request);
       }
     }
 
-    // Decide upon owners,
-    PRECICE_DEBUG("Decide owners, first round by rough load balancing");
-    PRECICE_ASSERT(ranksAtInterface != 0);
-    int localGuess = _mesh->getGlobalNumberOfVertices() / ranksAtInterface; // Guess for a decent load balancing
-    // First round: every slave gets localGuess vertices
-    for (int rank : utils::MasterSlave::allRanks()) {
-      int counter = 0;
-      for (size_t i = 0; i < slaveOwnerVecs[rank].size(); i++) {
-        // Vertex has no owner yet and rank could be owner
-        if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
-          slaveOwnerVecs[rank][i]                 = 1; // Now rank is owner
-          globalOwnerVec[slaveGlobalIDs[rank][i]] = 1; // Vertex now has owner
-          counter++;
-          if (counter == localGuess)
-            break;
-        }
+    for (auto &neighborRank : sharedVerticesSendMap) {
+      int receiveSize = 0;
+      utils::MasterSlave::_communication->receive(receiveSize, neighborRank.first);
+      if (receiveSize != 0) {
+        std::vector<double> receivedSharedVertices;
+        utils::MasterSlave::_communication->receive(receivedSharedVertices, neighborRank.first);
+        sharedVerticesReceiveMap.insert(std::make_pair(neighborRank.first, receivedSharedVertices));
       }
     }
 
-    // Second round: distribute all other vertices in a greedy way
-    PRECICE_DEBUG("Decide owners, second round in greedy way");
-    for (int rank : utils::MasterSlave::allRanks()) {
-      for (size_t i = 0; i < slaveOwnerVecs[rank].size(); i++) {
-        if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
-          slaveOwnerVecs[rank][i]                 = 1;
-          globalOwnerVec[slaveGlobalIDs[rank][i]] = rank + 1;
+    // wait until aSend is complete.
+    for (auto &rqst : vertexListRequests) {
+      rqst->wait();
+    }
+
+    // #6: Second round assignment according to the number of owned vertices
+
+    /* In case that a vertex can be shared between two ranks, the rank with lower
+       vertex count will own the vertex. 
+       If both ranks have same vertex count, the lower rank will own the vertex.
+    */
+
+    for (int i = 0; i < sharedVerticesGlobalIDs.size(); i++) {
+      bool owned = true;
+
+      for (auto &sharingRank : sharedVerticesReceiveMap) {
+        std::vector<double> vec = sharingRank.second;
+        if (std::find(vec.begin(), vec.end(), sharedVerticesGlobalIDs[i]) != vec.end()) {
+          if ((counter > neighborRanksVertexCount[sharingRank.first]) ||
+              (counter == neighborRanksVertexCount[sharingRank.first] && utils::MasterSlave::getRank() > sharingRank.first)) {
+            owned = false;
+          }
         }
       }
+      if (owned) {
+        tags[sharedVerticesLocalIDs[i]] = 1;
+      } else {
+        tags[sharedVerticesLocalIDs[i]] = 0;
+      }
```

Comment:
Can be written a bit more compact.

```suggestion
      tags[sharedVerticesLocalIDs[i]] = owned ? 1 : 0;
```